### ISSUE TITLE
Support to store column-major weights

### ISSUE THREAD

**Issue by freemin7** [huggingface/safetensors#367]
### Feature request

Define a way to store column-major weights in safe tensor in a standardised manner without requiring annotating files as row vs col when dealing with both.

Other implementations should either error when reading a file in an unexpected order, flip it in memory or execute it in that orientation (being possibly slower). Ideally without changes being required to these implementations. Other implementations must not produce wrong results because they operate on the matrix wrongly.

While row has won, col might still be worth supporting
https://github.com/huggingface/safetensors/commit/64464c91e940c87f8e0a150377db47c1fa2d0822

### Motivation

I am in possession of a NEC VectorEngine. An high memory bandwidth accelerator which best operates on column-major matrices. My usecase should be seen as a bespoke (and probably open source) executor of safetensor files which supports certain networks and i want to use safetensor as exchange format. 
![ve20b_llama2_c_plot](https://github.com/huggingface/safetensors/assets/25376460/67ce859a-1f73-42ed-a46d-a6e9284ced9a)
For more details see https://sx-aurora.github.io/posts/Llama2-on-VE-bf16/ in particular the performance evaluation section.

To efficiently run models i thus have to convert models between row and column order and have to deal with almost duplicate files which only differ in that detail which has to be tracked in the file name. Flipping the matrices in memory is wasteful in particular for short runs.

I would be delighted if i could use a standard safetensor tool to process both row and column order files and that i don't have to invent my own safetensor derived file format.

An useful generalisation would be to give the order of the axis of the higher order tensor stored in the file but that is not necessary for my use case so far.

This hardware due to it's simple programming paradigm is a nice playground for trying out new stuff.

### Your contribution

I don't immediately see a way how i can contribute to the effectively implementation as i am not from an ML background.

I could review and contribute to the design discussions.

Should there be a need to test on my hardware, i will give access liberally.

**Comment by Narsil** [comment]
Thanks for the detailed issue an motivation.

I don't think we should add this.

Having a clear layout makes everything much simpler to work with, and makes memory-mapping (zero-copy) predictable, which it wouldn't be if we supported col-major on platforms/engines that didn't work with it.

This is the same reason why endianness isn't there either.

When I first wrote safetensors, I was mostly preoccupied with safety (just don't execute code when loading tensors). Zero-copy and fast loading came afterwards, but it definitely seems like a major selling point (it seems to be your current issue for instance).
The issue with adding either flags is that is makes loading a safetensor file non-predictable. Can we zero-copy, or will we need to reorder (either shapes or endianness) ? No one can know without looking at the file, which seems bad imo.

The speed aspect you mention is really relevant, but the engine you're mentioning I had never heard of before so it seems a little bit non-obvious why we should break the predictable layout for a non standard platform. (For Cuda or Metal I would have a different answer for instance).

The good thing, is that for row-order, you can just save them in row-major with reversed shapes and just reinterpret the shapes on the fly on your particular platform. This would be non breaking, and allow you to have the speed back (I agree it's a bit cumbersome, but having specific files is necessary anyway to get col-major zero-copy, that doesn't change).

Another thing, is that for cuda, most of the time is dominated by transferring from disk to NVRAM anyway, therefore reordering bits is usually an acceptable cost, anyway you have benchmarked this for your platform ? (mmap is usually counterintuitive since it will return early, but load memory only on pagefaults, therefore hiding the loading cost somewhere else usually)

**Comment by freemin7** [comment]
> Another thing, is that for cuda, most of the time is dominated by transferring from disk to NVRAM anyway, therefore reordering bits is usually an acceptable cost, anyway you have benchmarked this for your platform ?

The NEC VE tends to have really good IO performance. But honestly i thought that Tensorflow probably just has a flag you need to set when loading/saving row-ordered vs col ordered and i rather trust that implementation to get that switch right then implement my own and have hard to find bugs. In particular implementing an inplace reorder seems fun until it isn't and one tears his hairs and i want to get started with baby steps.

> Having a clear layout makes everything much simpler to work with, and makes memory-mapping (zero-copy) predictable, which it wouldn't be if we supported col-major on platforms/engines that didn't work with it.

Your reasoning makes sense and i can reduce the scope of my inquiry a bit.

# New scope

Support for col-major doesn't become part of safetensor officially but instead i implement an extension to safetensor and only my implementation needs to handle both ways. I am looking for suggestions on how to design that extension.

# RFC

As i understand it, safetensor files not only store the data but also the computational graph. 
For 1d tensors no changes are needed.
In my understanding for 2d tensors the difference between row-major and col-major is the transpose operation. 

## Can we abuse the transpose operation?

My idea for my informal safetensor extension is that a -major stored matrix `tensor-name` is used referred to as `transpose(tensor-name)`  in the computational graph.
That means the original computational graph only gets to see data in row major data format. A col major interpreter would load all data with an additional unwritten transpose and would cancel double transposes.

# Points where i would like someone's input

- How to express that transpose idea idiomatically?
- Can this change be backwards compatible? 

A unmodified safetensor implementation reads a col-major matrix, transposes it and ends up with the right matrix for the computation but the computation might be slowed.

- How about higher order tensors in safetensor, is there some already supported syntax to express the generalisation of transpose for tensors somehow?

As i understand it the computational graph specification leans heavily on Pytorch or takes all of it outright. Even if some operations i describe here are not part of the safetensor standard, taking the same operation from pytorch might make it compatible by accident in the future if that operation is added.

# Why seek permission?

I am not familiar enough with the AI ecosystem and torch to tell whether my ideas are possible and before i commit a meaningful amount of time trying to get familiar i'd like to get sanity check that it is possible. This is the type of problem where a 5 min conversation with an expert gives a more definite result than 10 hours of reading docs as a person starting from basically zero. Some public documentation and buy-in prevents the wheel being reinvented by an unrelated party.

**Comment by freemin7** [comment]
I am still waiting for a comment.
